BeeJet.Bot/ClientHandlers/SlashCommandHandler.cs

```diff
@@ -1,32 +1,76 @@
+            return commandSources.SelectMany(c => c.GetMethods()
+                                                .Where(m =>
+                                                m.GetCustomAttribute<BeeJetBotSlashCommandAttribute>() != null
+                                                ).Select(m => (ClassType: c, Method: m, CommandName: m.GetCustomAttribute<BeeJetBotSlashCommandAttribute>().CommandName))).ToList();
+        }
+
+        public static IEnumerable<Type> GetCommandSourceTypes()
+        {
+            var type = typeof(ICommandSource);
+            return AppDomain.CurrentDomain.GetAssemblies()
+                               .SelectMany(s => s.GetTypes())
+                               .Where(p => type.IsAssignableFrom(p) && !p.IsAbstract);
         }
 
         internal async Task SlashCommandExecuted(SocketSlashCommand slashCommandArguments)
         {
-            var commandHandler = _commandSources.FirstOrDefault(b => b.GetCommandNames().Contains(slashCommandArguments.CommandName));
-            if (commandHandler != null)
+            var commandHandler = _commandMethods.FirstOrDefault(b => b.CommandName.Equals(slashCommandArguments.CommandName, StringComparison.OrdinalIgnoreCase));
+            if (commandHandler.ClassType != null)
+            {
+                var handlerInstance = _serviceProvider.GetService(commandHandler.ClassType);
+                if (handlerInstance != null)
+                {
+                    SlashCommandContext context = new SlashCommandContext(slashCommandArguments);
+                    await context.Initialize(_client);
+                    commandHandler.Method.Invoke(handlerInstance, new object[] { context });
+                }
+            }
+        }
+
+        internal async Task OnClientReadyAsync()
+        {
+            foreach (var commandSource in _commandMethods)
             {
-                var handler = commandHandler.GetCommandHandler(slashCommandArguments.CommandName, slashCommandArguments);
-                if (handler != null)
+                var attribute = commandSource.Method.GetCustomAttribute<BeeJetBotSlashCommandAttribute>();
+                foreach (var guild in _client.GetBotGuilds())
                 {
-                    await handler.Initialize(_client);
-                    await handler.SlashCommandExecuted();
+                    SlashCommandBuilder builder = new SlashCommandBuilder();
+                    var guildCommand = new SlashCommandBuilder();
+                    guildCommand.WithName(attribute.CommandName);
+                    guildCommand.WithDescription(attribute.Description);
+                    if (!string.IsNullOrWhiteSpace(attribute.BuilderMethod))
+                    {
+                        var builderMethod = commandSource.ClassType.GetMethods().FirstOrDefault(b => b.Name == attribute.BuilderMethod && b.GetParameters().Length == 1 && b.GetParameters()[0].ParameterType == typeof(SlashCommandBuilder));
```

Comment:
Dit gebeurt nu voor elke guild maar volgens mij is de `builderMethod` constant en alleen afhankelijk van de `commandSource`. Je hoeft hem hier alleen maar toe te passen voor elke guild waarin je het command toevoegt.

Ik denk dus dat je dit één foreach omhoog kan plaatsen :)

Reply:
Done